test_application/src/main.rs

```diff
@@ -0,0 +1,272 @@
+                // let key_gen = KeygenParticipant::new(config.id.clone(),
+                // config.other_ids.clone());
+
+                let participant: Participant<KeygenParticipant> =
+                    Participant::from_config(&config, sid, ());
+                let (public_key_shares_local, private_key_share_local) =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                public_key_shares = Some(public_key_shares_local);
+                private_key_share = Some(private_key_share_local);
+            }
+            SubProtocol::StartAuxInfo => {
+                let participant: Participant<AuxInfoParticipant> =
+                    Participant::from_config(&config, sid, ());
+                // TODO: Aux info takes no arguments?!
+                let (aux_info_public_local, aux_info_private_local) =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                aux_info_public = Some(aux_info_public_local);
+                aux_info_private = Some(aux_info_private_local);
+            }
+            SubProtocol::StartPresign => {
+                let input: PresignInput = PresignInput::new(
+                    aux_info_public.clone().unwrap(),
+                    aux_info_private.clone().unwrap(),
+                    public_key_shares.clone().unwrap(),
+                    private_key_share.clone().unwrap(),
+                )?;
+                let participant: Participant<PresignParticipant> =
+                    Participant::from_config(&config, sid, input);
+                let presign_record_local =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                presign_record = Some(presign_record_local);
+            }
+            SubProtocol::Sign => {
+                // TODO Weird that we create a vacuous participant for signing.
+                let signature =
+                    Participant::<KeygenParticipant>::sign(presign_record.take().unwrap(), hash())?;
+                info!("Computed signature: {:?}", signature);
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn hash() -> Sha256 {
+    let mut hasher = Sha256::new();
+    hasher.update(b"hello world");
+    hasher
+}
+
+#[instrument(skip_all)]
+fn worker_handle_subprotocol<P: ProtocolParticipant>(
+    mut participant: Participant<P>,
+    incoming: &Receiver<Message>,
+    outgoing: &Sender<MessageFromWorker>,
+) -> anyhow::Result<P::Output> {
+    let rng = &mut thread_rng();
+
+    debug!("Starting sub-protocol.");
+    let init_message = participant.initialize_message();
+
+    // TODO: Why does processing workers require rng?
```

Comment:
It is a bit awkward to pass the rng to every call to process_single_message. The fact it needs a rng seems like an impl detail.

Reply:
Continuing conversation on #290